src/basic/cap-list.c

```diff
@@ -64,3 +67,64 @@ int capability_from_name(const char *name) {
 int capability_list_length(void) {
         return (int) ELEMENTSOF(capability_names);
 }
+
+int capability_set_to_string_alloc(uint64_t set, char **s) {
+        _cleanup_free_ char *str = NULL;
+        const char *p;
+        unsigned long i;
+        size_t allocated = 0, n = 0;
+
+        assert(s);
+
+        for (i = 0; i < cap_last_cap(); i++)
+                if (set & (UINT64_C(1) << i)) {
+
+                        p = capability_to_name(i);
+                        if (!p)
+                                return -EINVAL;
+
+                        n += strlen(p) + 2;
+
+                        if (!GREEDY_REALLOC0(str, allocated, n))
+                                return -ENOMEM;
+
+                        strcat(str, p);
```

Comment:
not that it matters much but strcat searches for the end of the string first, and then writes the string there. But that's kinda unnecessary, we already know the location anyway, after all it's the old "n"... hence I'd change this to:

```
size_t add;

p = capability_to_name(i);
if (!p)
        return -EINVAL;

add = strlen(p);

if (GREEDY_REALLOC(str, allocated, n + add + 2))
        return -ENOMEM;

strcpy(mempcpy(str + n, p, add), " ");
n += add + 1;
…
```